### PR TITLE
Breaking up sky rendering into parts

### PR DESCRIPTION
**Pull request by Abastro**
This PR breaks sky rendering into parts of sky layers, so that each part can be replaced individually.
Each sky layer can be registered/replaced/removed individually by name.

Also, each sky layer has their own parent - the parent forms a group of sky layers, making it easy to group similar layers.

Order can be specified between any two elements, so that one element comes before another.

There are convenience methods like registerBefore/registerAfter to conveniently add a layer before/after a layer.

### CODE REVIEW

File: patches/minecraft/net/minecraft/world/WorldProvider.java.patch [hunk 1]
```diff
@@ -65,12 +65,13 @@
...
  
      @SideOnly(Side.CLIENT)
-@@ -241,6 +216,336 @@
+@@ -241,6 +216,352 @@
          return new WorldBorder();
      }
  
 +    /*======================================= Forge Start =========================================*/
 +    private net.minecraftforge.client.IRenderHandler skyRenderer = null;
++    private net.minecraftforge.client.SkyRenderer theSkyRenderer = new net.minecraftforge.client.SkyRenderer();
```
Comment (Gegy) [inline patches/minecraft/net/minecraft/world/WorldProvider.java.patch]: The similarity of the names `skyRenderer` and `theSkyRenderer` could be misleading. Should they not be renamed to make the difference more clear?

Comment (williewillus) [reply patches/minecraft/net/minecraft/world/WorldProvider.java.patch]: yeah, also `the` as a prefix is an antiquated MCP-ism, and really shouldn't be there :P

Comment (Abastro) [reply patches/minecraft/net/minecraft/world/WorldProvider.java.patch]: Of course, I'll fix that.

**Comment by Abastro** [comment]
Why isn't this getting any tags?
Besides, I really want some input from modders changing sky rendering. Anyone?

File: patches/minecraft/net/minecraft/world/WorldProvider.java.patch [hunk 2]
```diff
@@ -65,12 +65,13 @@
      }
  
      @SideOnly(Side.CLIENT)
-@@ -241,6 +216,336 @@
+@@ -241,6 +216,352 @@
          return new WorldBorder();
      }
  
 +    /*======================================= Forge Start =========================================*/
 +    private net.minecraftforge.client.IRenderHandler skyRenderer = null;
```
Comment (Gegy) [inline patches/minecraft/net/minecraft/world/WorldProvider.java.patch]: Could this be renamed to something like `legacySkyRenderer`? It's not public, so binary compatibility wouldn't be lost.

File: patches/minecraft/net/minecraft/world/WorldProvider.java.patch [hunk 3]
```diff
@@ -118,19 +119,34 @@
 +        return 1.0;
 +    }
 +
++    @Deprecated
```
Comment (Gegy) [inline patches/minecraft/net/minecraft/world/WorldProvider.java.patch]: Should add a note here why it is deprecated, and how the new system can be used instead.

File: patches/minecraft/net/minecraft/world/WorldProvider.java.patch [hunk 4]
```diff
@@ -118,19 +119,34 @@
...
 +
++    @SideOnly(Side.CLIENT)
++    public net.minecraftforge.client.SkyRenderer getSkyRenderHandler()
++    {
++        return this.skyRenderHandler;
++    }
++
++    /** Sets sky renderer for certain pass and resource location. Renderer with higher priority is called earlier. */
++    @SideOnly(Side.CLIENT)
++    public void setSkyRenderer(net.minecraftforge.client.EnumSkyPass pass, net.minecraft.util.ResourceLocation id, @Nullable net.minecraftforge.client.IRenderHandler renderer, double priority)
```
Comment (Gegy) [inline patches/minecraft/net/minecraft/world/WorldProvider.java.patch]: Should it rather be forced that `setSkyRenderer` should be called on the `SkyRenderer`? The method of this name implies that it's setting a field, and things will be overridden if you do it twice. (even with a different pass)
Or maybe a different name would be suitable? Maybe `setSkyPassRenderer` or `setSkyRendererPass`?

Edit: As you can have more than one pass renderer, `add` might be better than `set`

File: src/main/java/net/minecraftforge/client/EnumSkyPass.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.client;
+
+public enum EnumSkyPass
```
Comment (Gegy) [inline src/main/java/net/minecraftforge/client/EnumSkyPass.java]: From what I've seen, Forge has mostly been moving away from enums prefixed with `Enum`. Could be changed to `SkyRenderPass`

File: src/main/java/net/minecraftforge/client/SkyRenderer.java [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+    }
+
+    private static class RenderHandler implements Comparable<RenderHandler> {
+        IRenderHandler handler;
+        double priority;
+
+        RenderHandler(IRenderHandler handler, double priority)
+        {
+            this.handler = handler;
+            this.priority = priority;
```
Comment (Gegy) [inline src/main/java/net/minecraftforge/client/SkyRenderer.java]: Could priority be simplified to an enum of priority types, similar to how event priorities are handled?

File: src/main/java/net/minecraftforge/client/SkyRenderer.java [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+    private int glSkyList, glSkyList2;
+    private VertexBuffer skyVBO, sky2VBO;
+    private int starGLCallList;
+    private VertexBuffer starVBO;
+
+    private class SkyBackRenderer extends IRenderHandler {
+        @Override
+        public void render(float partialTicks, WorldClient world, Minecraft mc)
+        {
+            if (mc.world.provider.getDimensionType().getId() == 1)
```
Comment (Gegy) [inline src/main/java/net/minecraftforge/client/SkyRenderer.java]: These copied renderers could be hard to port if functionality changes in the original renderer. Would it be possible to call the vanilla renderer still, but filter what is rendered depending on the current pass?

File: patches/minecraft/net/minecraft/world/WorldProvider.java.patch (hunk 2 above)
Comment (Abastro) [reply patches/minecraft/net/minecraft/world/WorldProvider.java.patch]: I see.

File: patches/minecraft/net/minecraft/world/WorldProvider.java.patch (hunk 3 above)
Comment (Abastro) [reply patches/minecraft/net/minecraft/world/WorldProvider.java.patch]: Okay.

File: patches/minecraft/net/minecraft/world/WorldProvider.java.patch (hunk 4 above)
Comment (Abastro) [reply patches/minecraft/net/minecraft/world/WorldProvider.java.patch]: I see, moving to SkyRenderer would definitely be better.

File: src/main/java/net/minecraftforge/client/EnumSkyPass.java (hunk 5 above)
Comment (Abastro) [reply src/main/java/net/minecraftforge/client/EnumSkyPass.java]: I didn't know that, thanks!

File: src/main/java/net/minecraftforge/client/SkyRenderer.java (hunk 6 above)
Comment (Abastro) [reply src/main/java/net/minecraftforge/client/SkyRenderer.java]: No. It's priority like that applied on IWorldGenerator. It can't be replaced with enums because modders will want to add their own ordered renderers with arbitrary size.

File: src/main/java/net/minecraftforge/client/SkyRenderer.java (hunk 7 above)
Comment (Abastro) [reply src/main/java/net/minecraftforge/client/SkyRenderer.java]: Yes, it's possible. But I'll postpone the change till it's approved by modders doing sky rendering work. I guess this entire PR is moot before that.

**Comment by Abastro** [comment]
I need actual input from modder who's working with sky rendering.
Is this PR acceptable? Or is there a reason you can't use this way of modifying the sky rendering?

File: src/main/java/net/minecraftforge/client/SkyRenderer.java (hunk 6 above)
Comment (Gegy) [reply src/main/java/net/minecraftforge/client/SkyRenderer.java]: 👍

**Comment by HellFirePvP** [comment]
> I need actual input from modder who's working with sky rendering.

I can't seem to find a way to remove existing bits of skyrender? Removing a specific skyrender based on RenderPass and key would be _desperately_ needed. Other than that, i can't see anything wrong with it from my current perspective. On that note, this is looking way better than skyrendering works atm... :P

File: patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch [hunk 8]
```diff
@@ -93,29 +93,23 @@
...
      public void func_174976_a(float p_174976_1_, int p_174976_2_)
      {
-+        net.minecraftforge.client.IRenderHandler renderer = this.field_72769_h.field_73011_w.getSkyRenderer();
-+        if (renderer != null)
-+        {
-+            renderer.render(p_174976_1_, field_72769_h, field_72777_q);
-+            return;
-+        }
-+
++        net.minecraftforge.fml.client.FMLClientHandler.instance().renderSky(p_174976_1_);
```
Comment (covers1624) [inline patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch]: The return of this is never used, looks like the if got dropped when moving the logic.

Comment (Abastro) [reply patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch]: Oops.

**Comment by Abastro** [comment]
> I can't seem to find a way to remove existing bits of skyrender? Removing a specific skyrender based on RenderPass and key would be desperately needed.

I see, as renderers are already RenderPass-dependent I think I only have to add a method to unregister a sky renderer.

EDIT:
Now thinking about it, it's actully there; calling the register method with null renderer will remove the renderer. I think I need to special-case removing vanilla renderer, and it'll be enough.
Is it better with an unregister method?

Anyway, thanks for the feedback!

File: patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch (hunk 8 above)
Comment (Abastro) [reply patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch]: This is fixed now.

File: src/main/java/net/minecraftforge/client/SkyRenderHandler.java [hunk 9]
```diff
@@ -53,19 +55,33 @@
...
     }
 
+    /**
+     * Unregisters vanilla sky renderer for certain pass and id.
+     * This will only remove the renderer if it's the vanilla one.
+     * @param pass the sky render pass
+     * @param id the id to unregister
+     * @return the vanilla render handler, or <code>null</code> if there wasn't a vanilla renderer for specified pass-id pair
+     * */
+    public @Nullable IRenderHandler unregisterVanillaRenderer(SkyRenderPass pass, ResourceLocation id)
```
Comment (Gegy) [inline src/main/java/net/minecraftforge/client/SkyRenderHandler.java]: If there's only 1 vanilla renderer for each pass, does it make sense for the `id` to be passed in?

Comment (Abastro) [reply src/main/java/net/minecraftforge/client/SkyRenderHandler.java]: There are three vanilla renderers for SunMoonStars: the Sun, the Moon, and the Stars.

Comment (Gegy) [reply src/main/java/net/minecraftforge/client/SkyRenderHandler.java]: Ah, I see. Whoops.

**Comment by mcenderdragon** [comment]
This looks good, it will defanetly make it easier to add custom objects to modded dimensions too.

**Comment by Abastro** [comment]
Added some GL state calls for consistency when the vanilla renderers are removed.
Added test mods which either removes either of sun/moon/stars, or gives red tint on either sun or moon.
Completed this PR, and I think I got enough feedback.

Now, how can I get this PR be approved and pulled?

**Comment by Abastro** [comment]
Bump

**Comment by LexManos** [comment]
Do not bump.

**Comment by mezz** [comment]
If you are done requesting comments and ready to submit this as a feature PR, you can ping me and I will take a look. Otherwise I'll get around to it eventually as I look through all the PRs, like now.

At first look, please clean up the formatting. Patches should generally have braces on the same line to minimize patch lines, and non-patch code should always have braces on a new line.

Please update your original post to explain the PR as it is now, as a complete feature.

**Comment by Abastro** [comment]
Okay, @mezz. I fixed the formatting.

File: patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch [hunk 10]
```diff
@@ -93,29 +93,119 @@
...
-+        }
-+
++        int renderPass = net.minecraftforge.fml.client.FMLClientHandler.instance().renderSky(p_174976_1_);
++        if(renderPass == -1) return;
          if (this.field_72777_q.field_71441_e.field_73011_w.func_186058_p().func_186068_a() == 1)
          {
-             this.func_180448_r();
-@@ -1470,6 +1492,7 @@
+-            this.func_180448_r();
++            if(renderPass == 0 || renderPass == 1) this.func_180448_r();
```
Comment (mezz) [inline patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch]: You should put `if (renderPass == 0 || renderPass == 1)` on a separate line above `this.func_180448_r();` to keep this patch simple.

File: patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch [hunk 11]
```diff
@@ -93,29 +93,119 @@
...
+         }
+         else if (this.field_72777_q.field_71441_e.field_73011_w.func_76569_d())
+         {
+-            GlStateManager.func_179090_x();
++            if(renderPass == 0 || renderPass == 1) GlStateManager.func_179090_x();
```
Comment (mezz) [inline patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch]: put the `if` as a line above `GlStateManager.func_179090_x();`

File: patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch [hunk 12]
```diff
@@ -93,29 +93,119 @@
...
      public void func_174976_a(float p_174976_1_, int p_174976_2_)
      {
-+        net.minecraftforge.client.IRenderHandler renderer = this.field_72769_h.field_73011_w.getSkyRenderer();
-+        if (renderer != null)
-+        {
-+            renderer.render(p_174976_1_, field_72769_h, field_72777_q);
-+            return;
-+        }
-+
++        int renderPass = net.minecraftforge.fml.client.FMLClientHandler.instance().renderSky(p_174976_1_);
```
Comment (mezz) [inline patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch]: Can you use `SkyRenderPass` instead of `int` here? I think it would make things clearer.

File: patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch [hunk 13]
```diff
@@ -93,29 +93,119 @@
...
++            if(renderPass == 0 || renderPass == 1) GlStateManager.func_179090_x();
+             Vec3d vec3d = this.field_72769_h.func_72833_a(this.field_72777_q.func_175606_aa(), p_174976_1_);
+             float f = (float)vec3d.field_72450_a;
+             float f1 = (float)vec3d.field_72448_b;
+@@ -1275,9 +1292,10 @@
+                 f2 = f5;
+             }
+ 
+-            GlStateManager.func_179124_c(f, f1, f2);
++            if(renderPass == 0 || renderPass == 1) GlStateManager.func_179124_c(f, f1, f2);
```
Comment (mezz) [inline patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch]: `if` goes on newline before `GlStateManager.func_179124_c(f, f1, f2);`

File: patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch [hunk 14]
```diff
@@ -93,29 +93,119 @@
...
++            if(renderPass == 0 || renderPass == 1) {
+             GlStateManager.func_179132_a(false);
+             GlStateManager.func_179127_m();
+             GlStateManager.func_179124_c(f, f1, f2);
+@@ -1340,16 +1358,19 @@
+                 tessellator.func_78381_a();
+                 GlStateManager.func_179121_F();
+                 GlStateManager.func_179103_j(7424);
+-            }
++            }}
```
Comment (mezz) [inline patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch]: put your brace on a new line, leave the existing line alone

File: patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch [hunk 15]
```diff
@@ -93,29 +93,119 @@
...
+ 
++            if(renderPass == 0 || renderPass == 2) {
+             GlStateManager.func_179098_w();
+             GlStateManager.func_187428_a(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO);
+-            GlStateManager.func_179094_E();
++            GlStateManager.func_179094_E(); }
```
Comment (mezz) [inline patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch]: put your brace on a new line, leave the existing line alone

File: patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch [hunk 16]
```diff
@@ -93,29 +93,119 @@
...
+             float f16 = 1.0F - this.field_72769_h.func_72867_j(p_174976_1_);
++            if(renderPass == 0 || renderPass == 2) {
+             GlStateManager.func_179131_c(1.0F, 1.0F, 1.0F, f16);
+             GlStateManager.func_179114_b(-90.0F, 0.0F, 1.0F, 0.0F);
+-            GlStateManager.func_179114_b(this.field_72769_h.func_72826_c(p_174976_1_) * 360.0F, 1.0F, 0.0F, 0.0F);
++            GlStateManager.func_179114_b(this.field_72769_h.func_72826_c(p_174976_1_) * 360.0F, 1.0F, 0.0F, 0.0F); }
```
Comment (mezz) [inline patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch]: put your brace on a new line, leave the existing line alone

File: patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch [hunk 17]
```diff
@@ -93,29 +93,119 @@
...
+             float f17 = 30.0F;
++            if(renderPass == 0 || renderPass == 2) {
+             this.field_72770_i.func_110577_a(field_110928_i);
+             bufferbuilder.func_181668_a(7, DefaultVertexFormats.field_181707_g);
+             bufferbuilder.func_181662_b((double)(-f17), 100.0D, (double)(-f17)).func_187315_a(0.0D, 0.0D).func_181675_d();
+@@ -1357,7 +1378,15 @@
+             bufferbuilder.func_181662_b((double)f17, 100.0D, (double)f17).func_187315_a(1.0D, 1.0D).func_181675_d();
+             bufferbuilder.func_181662_b((double)(-f17), 100.0D, (double)f17).func_187315_a(0.0D, 1.0D).func_181675_d();
+             tessellator.func_78381_a();
++            GlStateManager.func_179121_F(); }
```
Comment (mezz) [inline patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch]: put your brace on a new line, leave the existing line alone

File: patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch [hunk 18]
```diff
@@ -93,29 +93,119 @@
...
++            GlStateManager.func_179114_b(-90.0F, 0.0F, 1.0F, 0.0F);
++            GlStateManager.func_179114_b(this.field_72769_h.func_72826_c(p_174976_1_) * 360.0F, 1.0F, 0.0F, 0.0F);
+             this.field_72770_i.func_110577_a(field_110927_h);
+             int k1 = this.field_72769_h.func_72853_d();
+             int i2 = k1 % 4;
+@@ -1372,6 +1401,12 @@
+             bufferbuilder.func_181662_b((double)f17, -100.0D, (double)(-f17)).func_187315_a((double)f22, (double)f23).func_181675_d();
+             bufferbuilder.func_181662_b((double)(-f17), -100.0D, (double)(-f17)).func_187315_a((double)f24, (double)f23).func_181675_d();
+             tessellator.func_78381_a();
++            GlStateManager.func_179121_F(); }
```
Comment (mezz) [inline patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch]: put your brace on a new line, leave the existing line alone

File: patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch [hunk 19]
```diff
@@ -93,29 +93,119 @@
...
++            GlStateManager.func_179114_b(-90.0F, 0.0F, 1.0F, 0.0F);
++            GlStateManager.func_179114_b(this.field_72769_h.func_72826_c(p_174976_1_) * 360.0F, 1.0F, 0.0F, 0.0F);
+             GlStateManager.func_179090_x();
+             float f15 = this.field_72769_h.func_72880_h(p_174976_1_) * f16;
+ 
+@@ -1393,12 +1428,13 @@
+                     GlStateManager.func_179148_o(this.field_72772_v);
+                 }
+             }
++            GlStateManager.func_179121_F(); }
```
Comment (mezz) [inline patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch]: put your brace on a new line, leave the existing line alone

File: patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch [hunk 20]
```diff
@@ -93,29 +93,119 @@
...
+-            GlStateManager.func_179121_F();
+             GlStateManager.func_179090_x();
+             GlStateManager.func_179124_c(0.0F, 0.0F, 0.0F);
+             double d3 = this.field_72777_q.field_71439_g.func_174824_e(p_174976_1_).field_72448_b - this.field_72769_h.func_72919_O();
+@@ -1464,12 +1500,13 @@
+             GlStateManager.func_179148_o(this.field_72781_x);
+             GlStateManager.func_179121_F();
+             GlStateManager.func_179098_w();
+-            GlStateManager.func_179132_a(true);
++            GlStateManager.func_179132_a(true); }
```
Comment (mezz) [inline patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch]: put your brace on a new line, leave the existing line alone

File: src/main/java/net/minecraftforge/client/SkyRenderHandler.java [hunk 21]
```diff
@@ -0,0 +1,217 @@
...
+        {
+            this.handler = handler;
+            this.priority = priority;
+        }
+
+        @Override
+        public int compareTo(RenderHandler o)
+        {
+            // Higher priority comes first
+            return -Comparator.<RenderHandler>comparingDouble(k -> k.priority).compare(this, o);
```
Comment (mezz) [inline src/main/java/net/minecraftforge/client/SkyRenderHandler.java]: `Comparator` has a `reversed()` method for this purpose

File: src/main/java/net/minecraftforge/client/SkyRenderPass.java [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+public enum SkyRenderPass
+{
+    /**
+     * Background sky.
+     * */
+    SKY_BACK,
+    /**
+     * Sun, moon, stars and others.
+     * */
+    SUN_MOON_STARS,
```
Comment (mezz) [inline src/main/java/net/minecraftforge/client/SkyRenderPass.java]: Why are these combined into one pass?
Is it possible to draw over the stars and behind the moon (like a planet would)?
Is it possible to make this into a more general sorted system, so that mods can draw before or after any specific part?

**Review by mezz — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch (hunk 12 above)
Comment (Abastro) [reply patches/minecraft/net/minecraft/client/renderer/RenderGlobal.java.patch]: I can't, because there are several vanilla renderers for 'SunMoonStars' pass.

File: src/main/java/net/minecraftforge/client/SkyRenderPass.java (hunk 22 above)
Comment (Abastro) [reply src/main/java/net/minecraftforge/client/SkyRenderPass.java]: I tried to keep what vanilla does. And I only got the three pass as they are necessary in most cases. While the Sun, the Moon and Stars are optional and even don't exist.
And yes, it's pretty tricky to draw over the stars and behind the moon in this system. I think I can solve this by making the priority be changed independently.
By sorted system, you mean something like mod sorting system? I think there is high probability where a cycle could happen. Also it's much more complex, so I don't think it's worth it.
EDIT: @mezz?

**Comment by Abastro** [comment]
@mezz, would you please check the replies?

**Review by mezz — APPROVED** [review]

**Comment by Abastro** [comment]
So this is how a PR is ignored? Is this basically an unacceptable PR?

**Comment by LexManos** [comment]
I typically leave my comments to private chats with Mezz as People think i'm to harsh. But hes on vacation and you insist on bumping your shit multiple times.
So speaking candidly, yes this PR is dumb, you split things but only partially and don't really split it to all the parts that it should be split into. Why are all the celestial objects in a single pass? is there any sorting system involved? 
Also some things are a bit confusing. You copy quite a bit of GLState code without actually telling what it does or where it comes from which causes issues with updating. You also use quite a bit of magic numbers, which is rather annoying.
Basically its a mess of confusing hack patches and considering I personally HATE rendering code i'm not inclined to clean it up for you. So i'm leaving it to Mezz, when he gets back.

**Comment by Abastro** [comment]
So what mezz said is what you want to say, right?

I think it could be better when it's just left unassigned - anyone could have said 'This PR is messy pile of junk, your poor reasoning is like that as well' and I'd have tried to rewrite this one without a question. (I know some people to help me out, so it should be much better)

Above all, please tell me if you think this concept itself is wrong. Maybe this is the real reason? No point in putting effort in this case.

**Comment by mezz** [comment]
Please update this PR to resolve conflicts, and run `gradle updateLicenses` to fix any new license headers.

**Comment by Abastro** [comment]
Resolved merge conflicts and draw a rough outline of SkyLayer system. How do you think?

**Comment by MajorTuvok** [comment]
You said that you want a proper sorting system - the topological sort currently used for sorting mods maybe?

**Comment by Abastro** [comment]
Yes, topological sort is what I'm going for.

**Comment by mezz** [comment]
Let me know when you are done updating this and I can take another look.

**Comment by Abastro** [comment]
@mezz I'm nearly finished updating it - only needs to work on the test code and performance now.
Would you review this PR once again? Now it revolves around tree of SkyLayer.
+ Discussed in discord, add registerBefore/registerAfter for convenience methods

**Comment by Abastro** [comment]
@mezz ? Please review

**Comment by mcenderdragon** [comment]
Just a quick performance question: are the stars worth cashing or is the performance vs memory effect too small ?

**Comment by Abastro** [comment]
@mcenderdragon Do you mean caching to VBO/display list? As far as I know, yes it is worth it - even though the effect is small.
I'll post a quick performance test when I finish the test.

**Comment by mcenderdragon** [comment]
yes the VBO/display list.

**Comment by liach** [comment]
Can you use the builtin topological sort library for the sky render layer list instead? Or is there any limitation with that library?

**Comment by Abastro** [comment]
@liach what's the name of the library?

**Comment by MajorTuvok** [comment]
It is not a library, it is part of forge (or fml to be precise).

The sole problem with it is, that the logs are hardcoded for sorting the modlist... 
Fixing that should not be difficult...

**Comment by Abastro** [comment]
@MajorTuvok The usecase is quite different, so I can't use that.

**Comment by MajorTuvok** [comment]
Why? The use cases are pretty similiar - you have sky layers which may depend on each other...

The sole effort is building the graph up correctly...

**Comment by Abastro** [comment]
But my one has hierarchy, and should preserve previous order if no order is specified for consistency with vanilla.

**Comment by Abastro** [comment]
@mezz It's really finished now, along with the test. The only thing left is performance test.
Please review this time.

**Comment by MajorTuvok** [comment]
Topological sorting is meant for cases where there are things dependant on each other - which is the case in a hierachy. 
And the default order will remain if specify your boundaries...

**Comment by Abastro** [comment]
Yeah, I mean the sky layer could have several sublayers. And I found the topological sort in the part too connected with the mod sorting stuffs. I can't use it.
And how do you retain the default order?

**Comment by MajorTuvok** [comment]
You can use it - f. e. adapting it to registries is possible and quite easy (I did that and then noticed that skyboy was already doing that and I then only applied the log fix to him).

You retain the default order by specifying it as boundaries. 
Give me 2 or 3 hours and I'll PR it to you - if I have the time to finish it, though I'll try.

**Comment by liach** [comment]
You can add a secondary sort to a topological sort when parallel entries exist. In this case, the secondary sort is the default order.

**Comment by MajorTuvok** [comment]
I just noticed that this is way more complex than I thought it is

**Comment by liach** [comment]
Can you move the content of the update runnable into a private method and use method reference instead?

**Comment by Abastro** [comment]
Yeah but I don't think the FML topology sort could make use of the secondary sort method easily.
And I'll switch to the method reference!

**Comment by liach** [comment]
Why can't you use a regular Layer to Layer map in favor to a graph for the layer group dependencies? One layer has at most one parent group (as I see)

**Comment by Abastro** [comment]
Because graph is more useful - it has many convenience methods

**Comment by MajorTuvok** [comment]
ok, so I have implemented the toposort now.

I'll test it now (and to do that I have to get IntelliJ to apply your existing patches! Argh!).

**Comment by MajorTuvok** [comment]
Something is really weird here - I can't start forge and trying to trick it by recompiling the Minecraft-codebase only works partially... somehow the SPacketEntityHeadLook and some others don't seem to exist. 
I think I did something wrong with the patches - I'll submit you a PR anyway.

**Comment by MajorTuvok** [comment]
If you want to, you can take a look at it @Abastro, even though I uploaded it, without being able to test it.

**Comment by liach** [comment]
IMO, a proper toposort for guava graph can be added. Probably one like this? 

https://github.com/KyoriPowered/lunar/blob/b5054bebfd98cb1f652645cf46dd51cb0771722d/src/main/java/net/kyori/lunar/graph/MoreGraphs.java#L72

**Comment by MajorTuvok** [comment]
That would be quite helpful, so that one doesn't always have to use the fml toposort which might very well be removed/replaced in 1.13, because of the rewrite. Though I'm not informed about that.

Besides I like this iterative (hope that's the correct translation) approach...
XD

You'd then only have to throw in your implementation @Abastro...?

**Comment by Abastro** [comment]
Closed an useless PR.

**Comment by MajorTuvok** [comment]
Why useless?
I do see quite a few uses...

**Comment by Daomephsta** [comment]
Yeah, this was shaping up nicely. It looked like it'd be a very versatile sky rendering system when it was done.

**Comment by Abastro** [comment]
Oops. many people care, actually.

**Comment by Abastro** [comment]
@liach How am I supposed to implement that? Do I need to copy the classes?

**Comment by Abastro** [comment]
Also fixed the description for the current state - I don't care about markdown format, btw

**Comment by Abastro** [comment]
@mezz I'm waiting for your feedbacks.

**Comment by MajorTuvok** [comment]
@Abastro, liach send you a link to the toposort...
Just send your Graph through it... (in your build method...)

**Comment by liach** [comment]
You can copy that code (it is mit licensed)

**Comment by liach** [comment]
moreover, i'd ask @kashike if he allows forge to take the whole sort code and license it under lgpl (without referring to his lunar library)

**Comment by liach** [comment]
rip.

**Comment by Abastro** [comment]
Oh no, how did you notice..
I wanted to close this silently =/

**Comment by Phylogeny** [comment]
@Abastro All 402 people watching this repo can get notifications when PRs/issues are opened/closed or when comments are added to them.

**Comment by liach** [comment]
the close was on forge discord github channel, too

**Comment by Abastro** [comment]
Sorry for bothering you again.
Anyway I lost the intent to continue this PR.
